src/Common/ThreadStatus.h

```diff
@@ -153,27 +161,32 @@ class ThreadStatus : public boost::noncopyable
     /// Use it only from current thread
     ContextWeakPtr query_context;
 
-    String query_id;
+    String query_id_from_query_context;
```

Comment:
It's interesting that `setQueryId` is called only from `MemoryTrackerThreadSwitcher`.
Maybe after we replace it with a thread group switcher, we can check if any `ThreadStatus` is created without a thread group but with query context (which is a bit strange). If not, we can probably move `query_id` (and query context) to the thread group.
(or even remove query_context and global_context from thread status finally)

Reply:
Ah, `attachQueryContext` is already removed. lol

Reply:
Yes. Now such invariant is vivid.
query_context could be removed from thread _group_. Other threads when attaching to the group copy query_context from it.
You might have meant `remove query_context  from thread status`? if so, I will consider this.
As I see, some time we copy some information from group to thread for the safety of signal listener code. That class took thread status at row pointer from an other thread and it has to be thread safe. The most difficult part in this part of code.